api/sessionrecording/session_recording.go

```diff
@@ -0,0 +1,287 @@
+package sessionrecording
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"io"
+	"log/slog"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/defaults"
+	apievents "github.com/gravitational/teleport/api/types/events"
+)
+
+const (
+	// Int32Size is a constant for 32 bit integer byte size
+	Int32Size = 4
+
+	// Int64Size is a constant for 64 bit integer byte size
+	Int64Size = 8
+
+	// MaxProtoMessageSizeBytes is maximum protobuf marshaled message size
+	MaxProtoMessageSizeBytes = 64 * 1024
+
+	// ProtoStreamV1 is a version of the binary protocol
+	ProtoStreamV1 = 1
+
+	// ProtoStreamV1PartHeaderSize is the size of the part of the protocol stream
+	// on disk format, it consists of
+	// * 8 bytes for the format version
+	// * 8 bytes for meaningful size of the part
+	// * 8 bytes for optional padding size at the end of the slice
+	ProtoStreamV1PartHeaderSize = Int64Size * 3
+
+	// ProtoStreamV1RecordHeaderSize is the size of the header
+	// of the record header, it consists of the record length
+	ProtoStreamV1RecordHeaderSize = Int32Size
+)
+
+// NewReader returns a new reader with slice pool
+func NewReader(r io.Reader) *Reader {
+	return &Reader{
+		reader:    r,
+		lastIndex: -1,
+	}
+}
+
+const (
+	// protoReaderStateInit is ready to start reading the next part
+	protoReaderStateInit = 0
+	// protoReaderStateCurrent will read the data from the current part
+	protoReaderStateCurrent = iota
+	// protoReaderStateEOF indicates that reader has completed reading
+	// all parts
+	protoReaderStateEOF = iota
+	// protoReaderStateError indicates that reader has reached internal
+	// error and should close
+	protoReaderStateError = iota
+)
+
+// Reader reads protobuf encoding from reader
+type Reader struct {
```

Comment:
Open to other names. It was suggested to avoid `ProtoReader` since its name leaked the implementation. I had this as `SessionRecordingReader`, but then it reads as a stutter when using, e.g. `sessionrecording.SessionRecordingReader` versus `sessionrecording.Reader`.

Reply:
`Reader` is a great name, since callers will refer to it as `sessionrecording.Reader`.

I would avoid mentioning protobuf in the godoc though.

Reply:
Good call, updated